components/validation/validation-custom-mixin.js

```diff
@@ -11,60 +11,49 @@ export const ValidationCustomMixin = superclass => class extends superclass {
 
 	constructor() {
 		super();
-		this._forElement = null;
+
+		this._eventSubscriberController = new EventSubscriberController(this, {},
+			{ eventName: 'd2l-validation-custom-connected' }
+		);
+
+		this._forPropertySubscriberController = new ForPropertySubscriberController(this,
+			{ onUnsubscribe: this._onUnsubscribe.bind(this) },
+			{ forProperty: 'for' }
+		);
 	}
 
 	get forElement() {
-		return this._forElement;
+		if (this.for) {
+			// Validation custom components only support one for element
+			return this._forPropertySubscriberController.providers[0];
+		} else {
+			return this._eventSubscriberController.provider;
+		}
 	}
 
 	connectedCallback() {
 		super.connectedCallback();
-		this._updateForElement();
-		this.dispatchEvent(new CustomEvent('d2l-validation-custom-connected', { bubbles: true }));
+		this._eventSubscriberController.hostConnected();
 	}
 
 	disconnectedCallback() {
 		super.disconnectedCallback();
-		if (isCustomFormElement(this._forElement)) {
-			this._forElement.validationCustomDisconnected(this);
-		}
-		this._forElement = null;
-		this.dispatchEvent(new CustomEvent('d2l-validation-custom-disconnected'));
+		this._eventSubscriberController.hostDisconnected();
+		this._forPropertySubscriberController.hostDisconnected();
 	}
 
 	updated(changedProperties) {
 		super.updated(changedProperties);
-
-		changedProperties.forEach((_, prop) => {
-			if (prop === 'for') {
-				this._updateForElement();
-			}
-		});
+		this._forPropertySubscriberController.hostUpdated(changedProperties);
 	}
 
 	async validate() {
 		throw new Error('ValidationCustomMixin requires validate to be overridden');
 	}
 
-	_updateForElement() {
-		const oldForElement = this._forElement;
-		if (this.for) {
-			const root = this.getRootNode();
-			this._forElement = root.getElementById(this.for);
-			if (!this._forElement) {
-				throw new Error(`validation-custom failed to find element with id ${this.for}`);
-			}
-		} else {
-			this._forElement = null;
-		}
-		if (this._forElement !== oldForElement) {
-			if (isCustomFormElement(oldForElement)) {
-				oldForElement.validationCustomDisconnected(this);
-			}
-			if (isCustomFormElement(this._forElement)) {
-				this._forElement.validationCustomConnected(this);
-			}
+	_onUnsubscribe() {
+		if (this._forPropertySubscriberController.provider.length === 0) {
+			throw new Error(`validation-custom failed to find element with id ${this.for}`);
```

Comment:
When does this end up getting thrown? Seems like before it would throw during `connectedCallback` -- which arguably could be too soon if there was a delay rendering the element it's looking for.

Having some kind of built-in error throwing capability in the controller itself might be a nice future enhancement too... could simplify the `LabelledBy` code as well as it does a lot of its own error throwing.

Reply:
Yeah I didn't test this out before/after but wanted to keep it in to remember - I assume it always "just worked" because people are putting the validator right beside the element.  I was looking at adding a "waitTime" to handle `LabelledBy` anyways, I wonder if I could generalize that to always search for three seconds, and then call an error callback.  I don't think we'd setup an error callback for selection or filter because it's not the end of the world if the element doesn't exist, they'll just handle themselves accordingly?

Reply:
I like the idea of an error callback a lot. 👍